File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.c
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/
// 0 ms 5.4 MB

#include <stdio.h>

struct ListNode {
    int val;
    struct ListNode* next;
};

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *current, *pioneer;
    current = pioneer = NULL;
    int count = 0;
    for (pioneer = head; pioneer && count <= n;
         pioneer = pioneer->next, count++);
    if (count <= n) {
        return head->next;
    }

    for (current = head; pioneer;
         current = current->next, pioneer = pioneer->next);
    current->next = current->next->next;

    return head;
}

File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.go
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/
// 2 ms 2.2 MB
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	pseudoHead := &ListNode{
		Next: head,
	}
	left := pseudoHead
	for i := 0; i < n+1; i++ {
		left = left.Next
	}
	right := pseudoHead
	for left != nil {
		left = left.Next
		right = right.Next
	}
	right.Next = right.Next.Next
	return pseudoHead.Next
}
